src/tap/src/tapcell.cpp

```diff
@@ -57,17 +63,1091 @@ Tapcell::~Tapcell()
+      }
+      std::pair<int, int> min_max_x;
+      min_max_x = getMinMaxX(rows); 
+    }
+  } else {
+     do_corners = 0;
+  }
+
+  int bottom_row = 0;
+  int top_row = rows.size()-1;
+
+  for (int cur_row = bottom_row;  cur_row <= top_row; cur_row++) {
+    for(odb::dbRow* subrow : rows[cur_row]) {
+      if (!(checkSymmetry(endcap_master, subrow->getOrient()))) {
+        continue;
+      }
+      odb::Rect row_bb;
+      subrow->getBBox(row_bb);
+      auto row_ori = subrow->getOrient();
+
+      int llx = row_bb.xMin();
+      int lly = row_bb.yMin();
+      int urx = row_bb.xMax();
+      int ury = row_bb.yMax();
+
+      if (do_corners) {
+        if (cur_row == top_row) {
+          masterl = pickCornerMaster(1, row_ori, cnrcap_nwin_master, cnrcap_nwout_master, endcap_master);
+          masterr = masterl;
+        } else if(cur_row == bottom_row) {
+          masterl = pickCornerMaster(-1, row_ori, cnrcap_nwin_master, cnrcap_nwout_master, endcap_master);
+          masterr = masterl;
+        } else {
+          auto rows_above = rows[cur_row + 1];
+          auto rows_below = rows[cur_row - 1];
+          masterl = pickCornerMaster(isXCorner(llx, rows_above, rows_below), row_ori, cnrcap_nwin_master, cnrcap_nwout_master, endcap_master);
+          masterr = pickCornerMaster(isXCorner(urx, rows_above, rows_below), row_ori, cnrcap_nwin_master, cnrcap_nwout_master, endcap_master);
+        }
+      } else {
+        masterl = endcap_master;
+        masterr = masterl;
+      }
+
+      if ((masterl->getWidth()) > (urx - llx)) {
+        continue;
+      }
+
+      buildCell(block, masterl, row_ori, llx, lly, prefix);
+
+      int master_x = masterr->getWidth();
+      int master_y = masterr->getHeight();
+
+      int loc_2_x = urx - master_x;
+      int loc_2_y = ury - master_y;
+      if (llx == loc_2_x && lly == loc_2_y) {
+        logger_->warn(utl::TAP, 9, "row {} have enough space for only one endcap.", subrow->getName());
+        continue;
+      }
+
+      odb::dbOrientType right_ori = row_ori;
+      if (row_ori == odb::dbOrientType::MX ) {
+        right_ori = odb::dbOrientType::R180;
+      } else if ( row_ori == odb::dbOrientType::R0) {
+        right_ori = odb::dbOrientType::MY;
+      }
+      buildCell(block, masterr, right_ori, loc_2_x, loc_2_y, prefix);
+    }
+  }
+
+  int endcap_count = Tapcell::phy_idx - start_phy_idx;
+  logger_->info(utl::TAP, 4, "Endcaps inserted: {}", endcap_count);
+  return endcap_count;
+}
+
+bool Tapcell::isXInRow(int& x, std::vector<odb::dbRow*>& subrow) {
+  for(odb::dbRow* row : subrow) {
+    odb::Rect row_bb;
+    row->getBBox(row_bb);
+    if (x >= row_bb.xMin() && x <= row_bb.xMax()) {
+      return 1;
+    }
+  }
+  return 0;
+}
+
+bool Tapcell::isXCorner(int& x, std::vector<odb::dbRow*>& rows_above, std::vector<odb::dbRow*>& rows_below) {
+  int in_above = isXInRow(x, rows_above);
+  int in_below = isXInRow(x, rows_below);
+
+  if (in_above && !in_below) {
+    return -1;
```

Comment:
The return type is bool so this makes no sense - it should be true/false or the function should return an int.